styx-api-service/src/main/java/com/spotify/styx/api/workflow/WorkflowInitializer.java

```diff
@@ -45,7 +45,8 @@ public WorkflowInitializer(Storage storage, Time time) {
     this.time = Objects.requireNonNull(time);
   }
 
-  public Optional<Workflow> inspectChange(Workflow workflow) {
+  public Optional<Workflow> inspectChangeAndStore(Workflow workflow)
+      throws WorkflowInitializationException {
     final Optional<Workflow> previous;
     try {
       previous = storage.workflow(workflow.id());
```

Comment:
Looking up the previous workflow, storing the new workflow and updating the trigger should be atomic, i.e. done in a transaction. Otherwise concurrent store requests might end up corrupting the workflow state.